simulation/behavior_tree_plugin/include/behavior_tree_plugin/pedestrian/behavior_tree.hpp

```diff
@@ -68,7 +68,7 @@ class PedestrianBehaviorTree : public BehaviorPluginBase
   DEFINE_GETTER_SETTER(StepTime,                                         double)
   DEFINE_GETTER_SETTER(TargetSpeed,                                      std::optional<double>)
   DEFINE_GETTER_SETTER(TrafficLightManager,                              std::shared_ptr<traffic_simulator::TrafficLightManager>)
-  DEFINE_GETTER_SETTER(UpdatedStatus,                                    std::shared_ptr<traffic_simulator::CanonicalizedEntityStatus>)
+  DEFINE_GETTER_SETTER(UpdatedStatus,                                    std::shared_ptr<traffic_simulator::EntityStatus>)
```

Comment:
Inside traffic_simulator, it is desirable to reduce exception handling by taking canonicalized LaneletPose/EntityStatus as arguments and return values whenever possible.
Therefore, please specify the circumstances under which this change must be resolved.

Reply:
@hakuturu583 
### Create CanonicalizedEntityStatus
In the approach proposed in this PR, the `CanonicalizedEntityStatus` object is created for a specific entity **only once** - during its spawn, more specifically here: [EntityManager::spawnEntity()](https://github.com/tier4/scenario_simulator_v2/blob/01f9907aa600be824d8e6230e7bf0c9da87acf92/simulation/traffic_simulator/include/traffic_simulator/entity/entity_manager.hpp#L473-L486). 
During the spawn, an `EntityStatus` object and a `CanonicalizedLaneletPose` object are created based on the arguments passed (parameters, pose etc.), these two objects forming the `CanonicalizedEntityStatus`. 
Created `CanonicalizedEntityStatus` is passed to the constructor of `EntityBase`. 

### Access to data in CanonicalizedEntityStatus
`EntityBase` has getters providing access to unmodifiable `CanonicalizedEntityStatus` data and setters allowing safe modification of selected objects. 
Especially it is impossible to modify `LaneletPose` itself. The only way to change the `CanonicalizedEntityStatus` (so also `LaneletPose`) is to pass the new `EntityStatus` to [`EntityBase::setStatus()`](https://github.com/tier4/scenario_simulator_v2/blob/01f9907aa600be824d8e6230e7bf0c9da87acf92/simulation/traffic_simulator/src/entity/entity_base.cpp#L532-L541).
This method calls [`CanonicalizedEntityStatus::set()`](https://github.com/tier4/scenario_simulator_v2/blob/01f9907aa600be824d8e6230e7bf0c9da87acf92/simulation/traffic_simulator/src/data_type/entity_status.cpp#L56-L91), where **canonicalization is performed** based on the passed `EntityStatus`. In such an architecture, it is not possible for `EntityBase` to have invalid `LaneletPose` (non-canonicalized) data.
Moreover, if the canonicalization fails, `CanonicalizedLaneletPose` simply will not have a value (it will be `std::nullopt`) - so the exception will not be thrown. 
Therefore, there is no need to handle exceptions in the case of canonicalization, only the `EntityBase::laneMatchingSucceed()` check is required - as it is developed for a long time in many parts of the code. 

### Advantages
The major advantage of this approach is that **during each tick** of the simulator, after the `BehaviorTree` updates the entity state (`EntityStatus`), canonicalization is performed **only once in one place**, to be more precise, for example, here: [VehicleEntity::onUpdate](https://github.com/tier4/scenario_simulator_v2/blob/01f9907aa600be824d8e6230e7bf0c9da87acf92/simulation/traffic_simulator/src/entity/vehicle_entity.cpp#L156).
Thanks to the fact that the canonicalization is performed in **one place**, on the existing `CanonicalizedEntityStatus`, it is always performed in the same way and in case `LaneletPose` is invalid, the objects needed for its setting are [available in advance](https://github.com/tier4/scenario_simulator_v2/blob/01f9907aa600be824d8e6230e7bf0c9da87acf92/simulation/traffic_simulator/src/data_type/entity_status.cpp#L79-L81): 
- `BoundingBox` from `CanonicalizedEntityStatus` is used, 
- `matching_distance` defined in `EntityBase::getDefaultMatchingDistanceForLaneletPoseCalculation()` is used,
- `EntityType` from `CanonicalizedEntityStatus` is used to deduce `include_crosswalk`.

Moreover, performing canonicalization using the previously existing `CanonicalizedEntityStatus`, helps to ensure that during the update some values of the `EntityStatus` [have not changed](https://github.com/tier4/scenario_simulator_v2/blob/01f9907aa600be824d8e6230e7bf0c9da87acf92/simulation/traffic_simulator/src/data_type/entity_status.cpp#L58-L63) (for example, the `BoundingBox` should be constant) - so it is **safe canonicalization**.

Thanks to the fact that canonicalization is performed in one place, we have a clear **division of responsibility** - from now on it is the full responsibility of `EntityBase` and its `EntityBase::setStatus()` methods to ensure that the data in `EntityBase->CanonicalizedEntityStatus->EntityStatus` are always correct (also canonicalized always in the same way).

## Circumstances
To achieve this (**safe canonicalization only once per tick in one place**), it is required to use `EntityStatus` as the result of an entity status update in `BehaviorTree`. Otherwise, canonicalization would be performed in `BehaviorTree` - in many places, in a way that is not always the same - in the implementation before this PR, there was a problem with `matching_distance`, which occurs in `EntityBase` and was not passed to `BehaviorTree`, which resulted in different values during canonicalization. 

Leaving the creation of `CanonicalizedEntityStatus` in `BehaviorTree` makes canonicalization unclear, can be unsafe, requires more resources, and it is unknown which module is in fact responsible for it.

To put it in other words, if `CanonicalizedEntityStatus` is created in other places, e.g. in `BehaviorTree`, the advantages from this PR will not be as great as it could be.

## Graph
Here I've attached a graph showing the architecture that this PR introduces.
![Canonicalization_description drawio](https://github.com/user-attachments/assets/e3e63f3b-c9fa-4241-a9f8-efaee9bba61e)

####  CanonicalizedLaneletPose  OpenScenarioInterpreter->API
As you can see in the graph there are 2 methods marked in red:
- `auto EntityBase::setCanonicalizedStatus(const CanonicalizedEntityStatus &) -> void;`  
- `auto CanonicalizedEntityStatus::set(const CanonicalizedEntityStatus & status) -> void;`

These are methods that cannot be eliminated at this time, since the communication between `OpenScenarioInterpreter` and the `API` is done via `CanonicalizedLaneletPose`, especially the [API::setEntityStatus()](https://github.com/tier4/scenario_simulator_v2/blob/01f9907aa600be824d8e6230e7bf0c9da87acf92/simulation/traffic_simulator/src/api/api.cpp#L107C1-L121) method accepts this data type - these are methods are used only for this purpose. 
This causes canonicalization to be performed on the `OpenScenarioInterpreter` side, which in my feeling should also be eliminated, and I believe it will not be difficult to achieve. 

To put it in other words, we should aim to make the canonicalization of `EntityStatus` happen in the [`EntityBase::setStatus()`](https://github.com/tier4/scenario_simulator_v2/blob/01f9907aa600be824d8e6230e7bf0c9da87acf92/simulation/traffic_simulator/src/entity/entity_base.cpp#L532-L541) method and only here.

Reply:
> EntityBase has getters providing access to unmodifiable CanonicalizedEntityStatus data members and setters allowing safe modification of selected data members.
Especially it is impossible to modify LaneletPose itself. The only way to change the CanonicalizedEntityStatus (so also LaneletPose) is to pass the new EntityStatus to [EntityBase::setStatus()](https://github.com/tier4/scenario_simulator_v2/blob/01f9907aa600be824d8e6230e7bf0c9da87acf92/simulation/traffic_simulator/src/entity/entity_base.cpp#L532-L541).
This method calls [CanonicalizedEntityStatus::set()](https://github.com/tier4/scenario_simulator_v2/blob/01f9907aa600be824d8e6230e7bf0c9da87acf92/simulation/traffic_simulator/src/data_type/entity_status.cpp#L56-L91), where canonicalization is performed based on the passed EntityStatus.
In such an architecture, it is not possible for EntityBase to have invalid LaneletPose (non-canonicalized) data.
Moreover, thanks to the fact that canonicalization takes place in one place, catching exceptions should be easier, especially since canonicalization is always done on the TrafficSimulator side.
It is worth noting that the only exception that can occur is in the case of lanelet_pose_valid == True and the trouble with canonicalization exactly [here](https://github.com/tier4/scenario_simulator_v2/blob/127bbaefb4d101e00828cd6fc6585cf76b9e03be/simulation/traffic_simulator/src/data_type/entity_status.cpp#L76) -> [here](https://github.com/tier4/scenario_simulator_v2/blob/127bbaefb4d101e00828cd6fc6585cf76b9e03be/simulation/traffic_simulator/src/utils/pose.cpp#L55) -> [here](https://github.com/tier4/scenario_simulator_v2/blob/127bbaefb4d101e00828cd6fc6585cf76b9e03be/simulation/traffic_simulator/src/data_type/lanelet_pose.cpp#L81-L88).

This is just the way it is currently implemented.
It is common in programming for assumptions to be unintentionally broken when adding functionality.
It is necessary to continue to ensure that types are canonicalized in order to reduce the risk of introducing bugs in the future.

Reply:
I think it is important to have one place to implement Canonicalize, but it would be better if it were just one function to Canonicalize.
If there are too many functions that take as arguments or return values LeneletPose/EntityStatus that have not been Canonicalized that may contain invalid values, additional error handling will be required in various places, which will increase maintenance costs.
Therefore, this implementation should be modified.

Reply:
@hakuturu583 

> This is just the way it is currently implemented.
> It is common in programming for assumptions to be unintentionally broken when adding functionality.
> It is necessary to continue to ensure that types are canonicalized in order to reduce the risk of introducing bugs in the future.

I fully agree that before this PR this was also guaranteed - in every place where such a problem could occur additional `CanonicalizedEntityStatus` objects were created - which solves the problem of invalid values, of course, but in my feelings it is not very efficient. I also want to point out the common occurrence in the code before PR of `static_cast<EntityStatus>(status)` casting and re-creating another `std::shared_ptr<CanonicalziedEntityStatus>` objects - which can be avoided without increasing the risk of invalid values - as developed in this PR.

Especially since, the methods from before this PR never operate on `CanonicalizedEntityStatus`, only in the line where the result is returned canonicalization takes place. In almost entire `BehaviorTree`, operations are performed in such a way that canonicalized input data are cast to `EntityStatus`, calculations are performed using `LaneletPose`, and canonicalization takes place only at return - once the object passed to `TrafficSimulator` is created - so at the end of each method performing the action.

https://github.com/tier4/scenario_simulator_v2/blob/39594fd362c91b32514863d4ee02b2d4dcaaf69a/simulation/behavior_tree_plugin/src/vehicle/lane_change_action.cpp#L263-L267

This PR only removes the creation of `CanonicalizedEntityStatus` in the return line of every method, but guarantees that this object is created in `EntityBase` after receiving the result from `BehaviorTree`.

I think that removing the creation of `CanonicalizedEntityStatus` in `BehaviorTree`, or indeed moving it to `EntityBase::setStatus()`, does not increase the chances of introducing bugs, but decreases them - since it guarantees that canonicalization will certainly always be performed before setting the status in `EntityBase`, and certainly in the already developed, uniform and proper way. 

> If there are too many functions that take as arguments or return values LeneletPose/EntityStatus that have not been Canonicalized that may contain invalid values, additional error handling will be required in various places, which will increase maintenance costs.

I believe that an approach in which `EntityStatus` and `LaneletPose` are allowed as input to `TrafficSimulator`, instead of their canonicalized equivalents will in fact protect us from the need to handle exceptions in multiple places. For example, every time `pose::canonicalize()` is called to return a result at the end of the methods in the `ActionNode` code, there is a risk of throwing an exception in those places, leading to the need to handle it.  Moving canonicalization to one place will allow catching errors in one place.

It is possible that I overlooked something and you mean a completely different error handling, if it is so please guide me. :bow: 

#### Summary

I fully agree that operations inside `TrafficSimulator` should be performed using `CanonicalizedLaneletPose` and `CanonicalizedEntityStatus` (as it is). I also agree that the input data for `BehaviorTree` should be canonicalized (as it is). However, I do not see any obstruction in making `TrafficSimulator` itself responsible for canonicalization of input data (also including those from `OpenScenarioInterpreter`). In other words, in my feelings, input data verification should be performed in `TrafficSimulator` itself. 

### Alternative solution

I've just developed an alternative solution and fully tested it - it works as expected. 

It is based on the fact that the communication between  `TrafficSimulator` and  `BehaviorTree` uses a shared_ptr. Therefore, nothing prevents `BehaviorTree` from writing data to the same `CanonicalizedEntityStatus` object that is in   `EntityBase`. To put it in other words - instead of returning the `EntityStatus` from `BehaviorTree` to `TrafficSimulator` it simply writes the result of the action directly to `CanonicaliedEntityStatus` - which is shared between `BehaviorTree` and `TrafficSimulator`. I think this solution is clear and efficient, but I'm not sure about its maintenance safety - I'm quite not a fan of using shared_ptr for such communication. 

Please let me know what you think about it. I've included the changes [in this summary commit](https://github.com/tier4/scenario_simulator_v2/pull/1320/commits/de4c942b8f2ce234b4a6bc95e457e7ed3df89558), it is quite large, but I hope it is readable. :bow: :bow: 

Example: [here](https://github.com/tier4/scenario_simulator_v2/blob/de4c942b8f2ce234b4a6bc95e457e7ed3df89558/simulation/behavior_tree_plugin/src/pedestrian/walk_straight_action.cpp#L50) -> [here](https://github.com/tier4/scenario_simulator_v2/blob/de4c942b8f2ce234b4a6bc95e457e7ed3df89558/simulation/behavior_tree_plugin/src/action_node.cpp#L97-L102) -> [here](https://github.com/tier4/scenario_simulator_v2/blob/de4c942b8f2ce234b4a6bc95e457e7ed3df89558/simulation/traffic_simulator/include/traffic_simulator/behavior/behavior_plugin_base.hpp#L58) -> [here](https://github.com/tier4/scenario_simulator_v2/blob/de4c942b8f2ce234b4a6bc95e457e7ed3df89558/simulation/traffic_simulator/include/traffic_simulator/entity/entity_base.hpp#L250) -> [here](https://github.com/tier4/scenario_simulator_v2/blob/de4c942b8f2ce234b4a6bc95e457e7ed3df89558/simulation/traffic_simulator/src/entity/pedestrian_entity.cpp#L257) -> [here](https://github.com/tier4/scenario_simulator_v2/blob/de4c942b8f2ce234b4a6bc95e457e7ed3df89558/simulation/traffic_simulator/src/entity/pedestrian_entity.cpp#L260-L261)

Reply:
> In other words, in my feelings, input data verification should be performed in TrafficSimulator itself.

I do not think so.
This is because the behavior plugin may plan in the lane coordinate system or in the map coordinate system, depending on the implementation of the behavior.
Instead of calculating and returning the “Next EntityStatus” inside the behavior plugin, it would be better to change the interface to “return the amount of movement” and do the lane coordinate system calculation inside the behavior plugin.

Reply:
```cpp
class Distance {
};

class LaneCoordinateDistance : public Distance {
};

class CartesiaonCoordinateDistance : public Distance {
};

class DeltaStatus
{
    std::unique_ptr<Distance> delta_distance;
    // traffic_simulator_msgs::msg::ActionStatus delta_action;
};

class CanonicalizedEntityStatus
{
    CanonicalizedEntityStatus operator+( DeltaStatus &other );
};

CanonicalizedEntityStatus CanonicalizedEntityStatus::operator+( DeltaStatus &other ){
    // Implement a function to calculate the next CanonicalizedEntityStatus considering DeltaStatus
};
```

I recommend expressing distances and coordinate systems as types as shown below.
If designed this way, there is no need to generate CanonicalizedEntityStatus many times, and the Canonicalize process will not be complicated by exception handling in both traffic_simulator/behavior_plugin.

Reply:
@hakuturu583 

> I do not think so.
> This is because the behavior plugin may plan in the lane coordinate system or in the map coordinate system, depending on the implementation of the behavior.

The effect of `BehaviorTree` action can indeed be reflected in the `EntityStatus` in one of the two coordinate systems, but we only consider two cases at the moment:
- the movement is planned in the map coordinate system (msg::Pose variable), the `lanelet_pose_valid` variable is set to **False** -  `lanelet_pose` variable is calculated based on pose with the canonicalized value inside `CanonicalizedEntityStatus::set()`
- movement is planned in the lanelet coordinate system (msg::LaneletPose variable), the `lanelet_pose_valid` variable is set to **True** - `lanelet_pose` variable is currently filled, only canonicalization is performed inside `CanonicalizedEntityStatus::set()` - here it is very essential that the pose variable (msg::Pose) also has to be filled in the `EntityStatus` received by `TrafficSimulator`

It can be seen here:
https://github.com/tier4/scenario_simulator_v2/blob/de4c942b8f2ce234b4a6bc95e457e7ed3df89558/simulation/traffic_simulator/src/data_type/entity_status.cpp#L66-L84

I feel that the above approach (from this PR) represents almost what you wrote using `DeltaStatus`, with one significant difference. Your idea handles the case when `lanelet_pose` is filled (`lanelet_pose_valid` variable is set to **True**), but `pose` (msg::Pose) is not filled - now such a case cannot occur and `pose` must always be filled. 
For this reason, I think your approach is really good and worth implementing, but I'm afraid not within this PR - it would introduce too many changes and possible troubles.

Could you please take a look at the [_Alternative Solution_ ](https://github.com/tier4/scenario_simulator_v2/pull/1320#discussion_r1695775781) presented above and tell me what you think about it?

Reply:
@hakuturu583 
Naturally, if the [Alternative Solution](https://github.com/tier4/scenario_simulator_v2/pull/1320#discussion_r1695775781) is not as expected, let me know :bow: :bow: 
 then I will revert the changes and bring back the creation of the `CanonicalizedEntityStatus` on the `BehaviorTree` side and sending it to the `TrafficSimulator` (as it was).

Reply:
> I feel that the above approach (from this PR) represents almost what you wrote using DeltaStatus, with one significant difference.

Hmmm, I don't feel that way.
I think the two advantages of my proposal are that the Canonicalize process is no longer needed inside the behavior plugin, and the C++ type can guarantee that the output from the behavior_plugin is valid.

This PR approach certainly eliminates the need for Canonicalize processing, but the second advantage disappears because it returns an uncanonicalized EntityStatus type.

Reply:
I see no problem with Alternative Solution as a temporary alternative.
However, the values exchanged between other behavior_plugins and traffic_simulator do not behave that way, so I would like to plan to reflect my proposal in the future.
However, I think that would be a big task and future work is fine.

Reply:
@dmoszynski If you finished this task, please talk to me on this thread.

Reply:
@hakuturu583 
The alternative solution is developed and has been tested locally. If you see a need for any additional changes to this PR, please let me know :bow:, if not, please allow me to run the regress tests again :bow:

Reply:
OK. I start reviewing.
Please wait a bit.